source/dhtslib/sam/cigar.d

```diff
@@ -24,13 +24,13 @@ struct Cigar
     /// Construct Cigar from raw data
     this(uint* cigar, uint length)
     {
-        ops = (cast(CigarOp*) cigar)[0 .. length];
+        ops = (cast(CigarOp*) cigar)[0 .. length].dup;
```

Comment:
Cigar stuff is likely called from hot loops; do you think this will have a negative impact on speed of client code? Certainly we have to address the segfault issue, but I wonder what other solutions may be possible? If segfault is due to GC, options could be limited. Same comment applies to the other constructor `.dup`

Reply:
R eplying to myself. If Dlang used a reference counting GC we wouldn't have this problem, if it is what I think it is (i.e. the read data to which the cigar data are pointing being freed)

Reply:
So this is to solve situations where the GC reaps the SAM record but the cigar is continuing to be used. This can happen if you're just doing cigar manipulation with the SAM record. GC is aggressive and doesn't wait for the SAM record to go out of scope since it is no longer being used. I certainly think the dup will affect performance negatively. I would prefer a solution that instead prevents reaping of SAM Record if there is still a reference to the backing bam1_t data. Maybe GC.addRange could help? I would like a better solution.

Reply:
Yes, GC.addRange or addRoot could help (although those calls might also slow things down, not sure about compared to a small string dup). Another solution is to disable GC in loops iterating over sam files which would speed things up evn irrespective of this cigar issue. Nevertheless, can you remind me when it might come up that one has a handle/ptr to the CIGAR data but have lost it to the backing SAMRecord? I am surprised that could happen

Reply:
```
foreach (SAMRecord read; bam.all_records)
{
    /// do something with record
    Cigar cigar = read.cigar;
    auto cigar_ops = cigar.ops; //cigar.ops is a slice to the original cigar data 
    
    /// SAMRecord no longer used
    /// SAMRecord reaped here
    /// do something with cigar_ops -> segfault
}
```

Since Cigar is just a wrapper around the backing cigar ops array, we previously just took a slice from the backing htslib bam variable data that contains the cigar ops. If the ```SAMRecord``` is reaped ever and the ```Cigar``` is not, the cigar's data is invalid. Hence the need for ```Cigar``` to have a copy of the cigar ops data.

Reply:
OK. Let me think about this for a couple of days. Probably over thinking it as many solutions may be as or more expensive than the .dup anyway.

Reply:
Though it is worth considering we could have this issue with any parts of dhtslib involving the sam record variable data, potentially our bam tags interface. I usually attempt to provide slices to existing data, rather than making copies for performance sake. We could potentially convert the variable length data of a bam record into proper slices so the GC has more control and notice. This could help the GC "see" when we make a slice to the variable length data and potentially provide more memory safety.